src/main/java/com/google/cloud/tools/ide/login/GoogleLoginState.java

```diff
@@ -429,46 +292,42 @@ private void retrieveSavedCredentials() {
       return;
     }
 
-    accessToken = savedAuthState.getAccessToken();
-    refreshToken = savedAuthState.getRefreshToken();
-    accessTokenExpiryTime = savedAuthState.getAccessTokenExpiryTime();
-    email = savedAuthState.getStoredEmail();
-    oAuth2Credential = makeCredential();
-
-    isLoggedIn = true;
+    Credential credential = makeCredential(savedAuthState.getAccessToken(),
+                                           savedAuthState.getRefreshToken());
+    String email = savedAuthState.getStoredEmail();
+    Account account = new Account(email, credential, savedAuthState.getAccessTokenExpiryTime());
+    accountRoster.addAccount(account);
   }
 
-  private void notifyLoginStatusChange(boolean login) {
+  private void notifyLoginStatusChange() {
+    Set<Account> accounts = listAccounts();  // Take a snapshot of accounts.
     synchronized(listeners) {
       for (LoginListener listener : listeners) {
-        listener.statusChanged(login);
+        listener.statusChanged(accounts);
       }
     }
   }
 
-  private String queryEmail() {
-    try {
-      HttpRequest get = createRequestFactory().buildGetRequest(new GenericUrl(GET_EMAIL_URL));
-      HttpResponse resp = get.execute();
-
-      String responseString = "";
-      try (Scanner scan = new Scanner(resp.getContent())) {
-        while (scan.hasNext()) {
-          responseString += scan.nextLine();
-        }
-      }
-
-      String userEmail = parseUrlParameters(responseString).get("email");
-      if (userEmail == null) {
-        loggerFacade.logWarning("Could not parse email after Google service sign-in");
+  @VisibleForTesting
+  String queryEmail(Credential credential) throws IOException, EmailAddressNotReturnedException {
+    HttpRequest get = transport.createRequestFactory(credential)
```

Comment:
I'm not sure how proxies figure into this, and how / if we tie it into Eclipse's proxy support.  But that's a bigger story and can wait.